tests/FSharp.Data.GraphQL.Tests/QueryWeightMiddlewareTests.fs

```diff
@@ -0,0 +1,365 @@
+    match result with
+    | Deferred(data, errors, deferred) ->
+        empty errors
+        data.["data"] |> equals (upcast expected)
+        deferred 
+        |> Observable.add (fun x -> 
+            actualDeferred.Add(x)
+            if actualDeferred.Count = 2 then mre.Set() |> ignore)
+        if TimeSpan.FromSeconds(float 30) |> mre.WaitOne |> not
+        then fail "Timeout while waiting for Deferred GQLResponse"
+        actualDeferred
+        |> Seq.cast<NameValueLookup>
+        |> contains expectedDeferred1
+        |> contains expectedDeferred2
+        |> ignore
+    | _ -> fail "Expected Deferred GQLResponse"
+    result.Metadata.TryFind<float>("queryWeightThreshold") |> equals (Some 2.0)
+
+[<Fact>]
+let ``Deferred and Streamed queries : Should not pass when above threshold``() =
+    let query = 
+        sprintf """query testQuery {
+            A (id : 1) {
+                id
+                value
+                subjects @%s {
+                    id
+                    value
+                    subjects {
+                        id
+                        value
+                    }
+                }                
+            }
+        }"""
+    asts query
+    |> Seq.map execute
+    |> Seq.iter (fun result ->
+        match result with
+        | Direct(data, errors) ->
+            errors |> equals expectedErrors
+            data.["data"] |> isNameValueDict |> empty
+        | _ -> fail "Expected Direct GQLResponse"
+        result.Metadata.TryFind<float>("queryWeightThreshold") |> equals (Some 2.0))
+
+[<Fact>]
+let ``Inline fragment query : Should pass when below threshold``() =
+    let query = 
+        parse """query testQuery {
+            A (id : 1) {
+                id
+                value
+                subjects {
+                    ... on A {
+                        id
+                        value
+                    }
+                    ... on B {
+                        id
+                    }
+                }                
+            }
+        }"""
+    let expected = 
+        NameValueLookup.ofList [
+            "A", upcast NameValueLookup.ofList [
+                "id", upcast 1
+                "value", upcast "A1"
+                "subjects", upcast [ 
+                    NameValueLookup.ofList [
+                        "id", upcast 2
+                        "value", upcast "A2"                    
+                    ]
+                    NameValueLookup.ofList [
+                        "id", upcast 6
+                    ]
+                ]
+            ]
+        ]
+    let result = execute query
+    match result with
+    | Direct (data, errors) ->
+        empty errors
+        data.["data"] |> equals (upcast expected)
+    | _ -> fail "Expected Direct GQLResponse"
+    result.Metadata.TryFind<float>("queryWeightThreshold") |> equals (Some 2.0)
+
+[<Fact>]
+let ``Inline fragment query : Should not pass when above threshold``() =
```

Comment:
If we use fragments though, and specify the field in only one of them, then we have a weight of 1 for it. In this case, the query still fails to execute because intentionally we queried a nested subjects on this fragment.